sw/device/tests/otbn_mem_scramble_test.c

```diff
@@ -22,19 +22,59 @@ typedef dif_result_t (*otbn_write_t)(const dif_otbn_t *otbn,
                                      uint32_t offset_bytes, const void *src,
                                      size_t len_bytes);
 
-/**
- * Number of distinct addresses to check in each IMEM and DMEM.
- */
-static const int kNumAddrs = 50;
-
-/**
- * Minimum number of expected integrity errors in IMEM and DMEM after
- * re-scrambling.
- *
- * Note that there is a non-zero chance for every word that the integrity bits
- * after re-scrambling are still valid.
- */
-static const int kNumIntgErrorsThreshold = kNumAddrs - 1;
+enum {
+  /**
+   * Number of distinct addresses to check in each IMEM and DMEM.
+   */
+  kNumAddrs = 50,
+
+  /**
+   * Minimum number of expected integrity errors in IMEM and DMEM after
+   * re-scrambling.
+   *
+   * Note that each word has 7 bits of ECC, so the probability that a word would
+   * generate an ECC error is (127/128). Then the probability that all
+   * `kNumAddrs`  will generate ECC errors is `pow(127/128, kNumAddrs)` after
+   * re-scrambling.
+   *
+   * The Generic formula:
+   *               (w-i)
+   *             127
+   * Pr(i) =  -------- x (w choose i)
+   *                w
+   *             128
+   * Where:
+   *      w = The number of words tested.
+   *      i = The number of words that may not generate errors.
+   *      Pr(i) = Probability that i words will not generate an ECC error.
+   *
+   * So for i in (0..4):
+   *
+   * ``` Python
+   * from math import comb
+   * w = 50
+   * t = 0
+   * for i in range(5):
+   *    p = ((127**(w-i))/(128**w)) * comb(w,i)
+   *    t += p
+   *    print(f'Pr({i}): { round(p, 4)},\tsum{{Pr(0-{i})}}: {round(t, 6)}')
+   * ```
+   * ```
+   * Pr(0): 0.6756,   sum{Pr(0-0)}: 0.675597
+   * Pr(1): 0.266,    sum{Pr(0-1)}: 0.94158
+   * Pr(2): 0.0513,   sum{Pr(0-2)}: 0.992891
+   * Pr(3): 0.0065,   sum{Pr(0-3)}: 0.999356
+   * Pr(4): 0.0006,   sum{Pr(0-4)}: 0.999954
+   * ```
+   * So by choosen `48 = (kNumAddrs - 2)` as the threshold will have the
+   * probability of `1 - 0.992891 = 0.71%` that this test will fail randomly due
+   * to ECC errors not generated. Which seems a reasonable number.
```

Comment:
If we have several tests like this, a 0.71% probability of failure for each test can still lead to a flaky CI. Ideally, we'd make that number lower, but I'm not sure how feasible that is. In some cases there are also ways to work around the problem like using a fixed seed (or set of seeds) for the tests.

Reply:
I don't see a problem in making this number bigger and make the test more robust, I just proposed theses numbers because I was thinking about the nightly regression that runs once a day, which would probably fails once a year moreless. But if we start to run these tests on CI, 0.71% would be anoing.